{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### parse data for vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def fix_transcript(row):\n",
    "    return row['Transcript'].split('.')[0]\n",
    "\n",
    "def choose_one(rows):\n",
    "    if len(rows) == 1:\n",
    "        return rows\n",
    "    else:\n",
    "        r = rows[rows.Classification != 'VUS']\n",
    "        if len(r) == 1:\n",
    "            return r\n",
    "        else:\n",
    "            return rows.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     chr     p.      c. Genomic Pos (hg19)  Ref (single-base)  \\\n",
      "0  chr22  A206S   c.616               12998                 G   \n",
      "1  chr22  A260V   c.779               14480                 C   \n",
      "2  chr22    A2V     c.5                 564                 C   \n",
      "3  chr22  A480G  c.1439               20507                 C   \n",
      "4  chr22  C483R  c.1447               20515                 T   \n",
      "\n",
      "  Alt (single-base)  Ref  Alt Exon (Gene) Gene Symbol test/Indication  \\\n",
      "0                 T  GCC  TCC     exn5-39        ADSL        Comp_epi   \n",
      "1                 T  GCA  GTA     exn7-14        ADSL        Comp_epi   \n",
      "2                 T  GCG  GTG     exn1+64        ADSL      Infant_epi   \n",
      "3                 G  GCA  GGA    exn13-68        ADSL      Infant_epi   \n",
      "4                 C  TGT  CGT    exn13-60        ADSL       Child_epi   \n",
      "\n",
      "    Transcript Classification  Pos Fam Cnt  Neg Fam Cnt  Homozygous Fam Cnt  \\\n",
      "0  NM_000026.2            VUS            8         8275                   0   \n",
      "1  NM_000026.2            VUS            1         8282                   0   \n",
      "2  NM_000026.2     PATHOGENIC            1         8282                   0   \n",
      "3  NM_000026.2            VUS            1         8282                   0   \n",
      "4  NM_000026.2            VUS            1         8282                   0   \n",
      "\n",
      "   Hemizygous Fam Cnt  Heterozygous Fam Cnt  \n",
      "0                   0                     8  \n",
      "1                   0                     1  \n",
      "2                   0                     1  \n",
      "3                   0                     1  \n",
      "4                   0                     1  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.4/site-packages/pandas/core/indexing.py:297: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self.obj[key] = _infer_fill_value(value)\n",
      "/opt/conda/lib/python3.4/site-packages/pandas/core/indexing.py:477: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self.obj[item] = s\n",
      "/opt/conda/lib/python3.4/site-packages/pandas/indexes/api.py:71: RuntimeWarning: unorderable types: int() < str(), sort order is undefined for incomparable objects\n",
      "  result = result.union(other)\n",
      "/opt/conda/lib/python3.4/site-packages/pandas/indexes/api.py:71: RuntimeWarning: unorderable types: str() > int(), sort order is undefined for incomparable objects\n",
      "  result = result.union(other)\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'chrom'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/opt/conda/lib/python3.4/site-packages/pandas/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2133\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2134\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2135\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/index.pyx\u001b[0m in \u001b[0;36mpandas.index.IndexEngine.get_loc (pandas/index.c:4433)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/index.pyx\u001b[0m in \u001b[0;36mpandas.index.IndexEngine.get_loc (pandas/index.c:4279)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/src/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas.hashtable.PyObjectHashTable.get_item (pandas/hashtable.c:13742)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/src/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas.hashtable.PyObjectHashTable.get_item (pandas/hashtable.c:13696)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'chrom'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-613b4297c6de>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_fix\u001b[0m\u001b[0;34m[\u001b[0m \u001b[0mdf_fix\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Ref (single-base)'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'no alleles found'\u001b[0m \u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0mg_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'chrom'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'pos'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'ref'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'alt'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mg_cols\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/conda/lib/python3.4/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mgroupby\u001b[0;34m(self, by, axis, level, as_index, sort, group_keys, squeeze, **kwargs)\u001b[0m\n\u001b[1;32m   3989\u001b[0m         return groupby(self, by=by, axis=axis, level=level, as_index=as_index,\n\u001b[1;32m   3990\u001b[0m                        \u001b[0msort\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgroup_keys\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgroup_keys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msqueeze\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msqueeze\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3991\u001b[0;31m                        **kwargs)\n\u001b[0m\u001b[1;32m   3992\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3993\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0masfreq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfreq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnormalize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.4/site-packages/pandas/core/groupby.py\u001b[0m in \u001b[0;36mgroupby\u001b[0;34m(obj, by, **kwds)\u001b[0m\n\u001b[1;32m   1509\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'invalid type: %s'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1510\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1511\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mklass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mby\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1512\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1513\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.4/site-packages/pandas/core/groupby.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, obj, keys, axis, level, grouper, exclusions, selection, as_index, sort, group_keys, squeeze, **kwargs)\u001b[0m\n\u001b[1;32m    368\u001b[0m                                                     \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    369\u001b[0m                                                     \u001b[0msort\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msort\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 370\u001b[0;31m                                                     mutated=self.mutated)\n\u001b[0m\u001b[1;32m    371\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    372\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.4/site-packages/pandas/core/groupby.py\u001b[0m in \u001b[0;36m_get_grouper\u001b[0;34m(obj, key, axis, level, sort, mutated)\u001b[0m\n\u001b[1;32m   2460\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2461\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mis_in_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# df.groupby('name')\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2462\u001b[0;31m             \u001b[0min_axis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgpr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2463\u001b[0m             \u001b[0mexclusions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2464\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgpr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mGrouper\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mgpr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkey\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.4/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2057\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2058\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2059\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2060\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2061\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_getitem_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.4/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_getitem_column\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2064\u001b[0m         \u001b[0;31m# get column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2065\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_unique\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2066\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2067\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2068\u001b[0m         \u001b[0;31m# duplicate columns & possible reduce dimensionality\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.4/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_get_item_cache\u001b[0;34m(self, item)\u001b[0m\n\u001b[1;32m   1384\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcache\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1385\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mres\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1386\u001b[0;31m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1387\u001b[0m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_box_item_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1388\u001b[0m             \u001b[0mcache\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.4/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, item, fastpath)\u001b[0m\n\u001b[1;32m   3541\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3542\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misnull\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3543\u001b[0;31m                 \u001b[0mloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3544\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3545\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0misnull\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.4/site-packages/pandas/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2134\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2135\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2136\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_maybe_cast_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2137\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2138\u001b[0m         \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtolerance\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtolerance\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/index.pyx\u001b[0m in \u001b[0;36mpandas.index.IndexEngine.get_loc (pandas/index.c:4433)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/index.pyx\u001b[0m in \u001b[0;36mpandas.index.IndexEngine.get_loc (pandas/index.c:4279)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/src/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas.hashtable.PyObjectHashTable.get_item (pandas/hashtable.c:13742)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/src/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas.hashtable.PyObjectHashTable.get_item (pandas/hashtable.c:13696)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'chrom'"
     ]
    }
   ],
   "source": [
    "dat_file = '../data/raw/EPIv6.xlsx'\n",
    "df_init = pandas.read_excel(dat_file)\n",
    "crit = df_init.apply(lambda row: not pandas.isnull(row['Transcript']), axis=1)\n",
    "df_pre = df_init[crit]\n",
    "print( df_pre.head() )\n",
    "\n",
    "# rm duplicate rows\n",
    "df_pre.loc[:, 'simple_nm'] = df_pre.apply(fix_transcript, axis=1)\n",
    "cols = [x for x in df_pre.columns.values\n",
    "        if x != 'Transcript']\n",
    "idx_vals = df_pre[cols].drop_duplicates().index.values\n",
    "df_fix = df_pre.loc[idx_vals][df_init.columns.values]\n",
    "\n",
    "# choose non-VUS row for duplicates\n",
    "g_cols = [x for x in df_fix.columns.values\n",
    "          if not x in ('Transcript', 'Classification')]\n",
    "df_fix.groupby(g_cols).apply(choose_one)\n",
    "df = df_fix[ df_fix['Ref (single-base)'] != 'no alleles found' ]\n",
    "g_cols = ('chr', 'pos', 'ref', 'alt')\n",
    "df.groupby(g_cols).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/perry/miniconda3/envs/mahdi_epi/lib/python3.6/site-packages/pandas/core/indexing.py:517: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self.obj[item] = s\n"
     ]
    }
   ],
   "source": [
    "def fix_ref(row):\n",
    "    ref = row['Ref (single-base)']\n",
    "    if ',' in ref:\n",
    "        return ref.split(',')[-1].split('>')[0]\n",
    "    return ref\n",
    "\n",
    "def fix_alt(row):\n",
    "    ref = row['Ref (single-base)']\n",
    "    if ',' in ref:\n",
    "        return ref.split(',')[-1].split('>')[1]\n",
    "    return row['Alt (single-base)']\n",
    "\n",
    "def mk_vcf_line(row, fout):\n",
    "    ls = (row['clinical_class'], row['pos_fam_count'], row['neg_fam_count'])\n",
    "    info = 'CLIN_CLASS=%s;POS_FAM_COUNT=%d;NEG_FAM_COUNT=%d' % ls\n",
    "    ls  = (row['chrom'], str(row['pos']), '.',\n",
    "           row['ref'], row['alt'], '.', '.',\n",
    "           info)\n",
    "    print('\\t'.join(ls), file=fout)\n",
    "\n",
    "def write_vcf(df, vcf_out):\n",
    "    with open(vcf_out, 'w') as fout:\n",
    "        print('##fileformat=VCFv4.2', file=fout)\n",
    "        print('##INFO=<ID=CLIN_CLASS,Number=1,Type=String,Description=\"pos_fam_count\">', file=fout)\n",
    "        print('##INFO=<ID=POS_FAM_COUNT,Number=1,Type=Integer,Description=\"pos_fam_count\">', file=fout)\n",
    "        print('##INFO=<ID=NEG_FAM_COUNT,Number=1,Type=Integer,Description=\"pos_fam_count\">', file=fout)\n",
    "        print('#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO', file=fout)        \n",
    "        df.apply(lambda row: mk_vcf_line(row, fout), axis=1)\n",
    "        \n",
    "df.loc[:, 'ref'] = df.apply(fix_ref, axis=1)\n",
    "df.loc[:, 'alt'] = df.apply(fix_alt, axis=1)\n",
    "df.loc[:, 'chrom'] = df.apply(lambda row: row['chr'][3:], axis=1)\n",
    "uc = cols = ['chrom', 'Genomic Pos (hg19) ', 'ref', 'alt',\n",
    "             'Classification', 'Pos Fam Cnt', 'Neg Fam Cnt', 'Homozygous Fam Cnt',\n",
    "             'Hemizygous Fam Cnt', 'Heterozygous Fam Cnt']\n",
    "new_cols = {'Genomic Pos (hg19) ': 'pos',\n",
    "            'Classification':'clinical_class',\n",
    "            'Pos Fam Cnt':'pos_fam_count',\n",
    "            'Neg Fam Cnt':'neg_fam_count',\n",
    "            'Homozygous Fam Cnt':'hom_fam_count',\n",
    "            'Hemizygous Fam Cnt':'hemi_fam_count',\n",
    "            'Heterozygous Fam Cnt':'het_fam_count'\n",
    "           }\n",
    "df_final = df[uc].rename(columns=new_cols).sort_values(by=['chrom', 'pos'])\n",
    "vcf_out = '../data/interim/EPIv6.vcf'\n",
    "write_vcf(df_final, vcf_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>clinical_class</th>\n",
       "      <th>pos_fam_count</th>\n",
       "      <th>neg_fam_count</th>\n",
       "      <th>hom_fam_count</th>\n",
       "      <th>hemi_fam_count</th>\n",
       "      <th>het_fam_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>12998</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>VUS</td>\n",
       "      <td>8</td>\n",
       "      <td>8275</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>14480</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>VUS</td>\n",
       "      <td>1</td>\n",
       "      <td>8282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>564</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>PATHOGENIC</td>\n",
       "      <td>1</td>\n",
       "      <td>8282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>20507</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>VUS</td>\n",
       "      <td>1</td>\n",
       "      <td>8282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>20515</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>VUS</td>\n",
       "      <td>1</td>\n",
       "      <td>8282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    pos ref alt clinical_class  pos_fam_count  neg_fam_count  \\\n",
       "0    22  12998   G   G            VUS              8           8275   \n",
       "1    22  14480   C   C            VUS              1           8282   \n",
       "2    22    564   C   C     PATHOGENIC              1           8282   \n",
       "3    22  20507   C   C            VUS              1           8282   \n",
       "4    22  20515   T   T            VUS              1           8282   \n",
       "\n",
       "   hom_fam_count  hemi_fam_count  het_fam_count  \n",
       "0              0               0              8  \n",
       "1              0               0              1  \n",
       "2              0               0              1  \n",
       "3              0               0              1  \n",
       "4              0               0              1  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>p.</th>\n",
       "      <th>c.</th>\n",
       "      <th>Genomic Pos (hg19)</th>\n",
       "      <th>Ref (single-base)</th>\n",
       "      <th>Alt (single-base)</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Exon (Gene)</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>...</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Pos Fam Cnt</th>\n",
       "      <th>Neg Fam Cnt</th>\n",
       "      <th>Homozygous Fam Cnt</th>\n",
       "      <th>Hemizygous Fam Cnt</th>\n",
       "      <th>Heterozygous Fam Cnt</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2542</th>\n",
       "      <td>chr13</td>\n",
       "      <td>M1?</td>\n",
       "      <td>c.2</td>\n",
       "      <td>1435</td>\n",
       "      <td>no alleles found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NNN</td>\n",
       "      <td>NNN</td>\n",
       "      <td>exn1+30</td>\n",
       "      <td>CLN5</td>\n",
       "      <td>...</td>\n",
       "      <td>NM_006493.2</td>\n",
       "      <td>PATHOGENIC</td>\n",
       "      <td>10</td>\n",
       "      <td>8535</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>no alleles found</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4033</th>\n",
       "      <td>chr13</td>\n",
       "      <td>Met1?</td>\n",
       "      <td>c.2</td>\n",
       "      <td>1435</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLN5</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>VUS</td>\n",
       "      <td>1</td>\n",
       "      <td>8544</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr      p.   c. Genomic Pos (hg19)  Ref (single-base)  \\\n",
       "2542  chr13     M1?  c.2                1435  no alleles found   \n",
       "4033  chr13  Met1?   c.2                1435                 T   \n",
       "\n",
       "     Alt (single-base)  Ref  Alt Exon (Gene) Gene Symbol  ...    Transcript  \\\n",
       "2542               NaN  NNN  NNN     exn1+30        CLN5  ...   NM_006493.2   \n",
       "4033                 C    T    C         NaN        CLN5  ...           NaN   \n",
       "\n",
       "     Classification Pos Fam Cnt  Neg Fam Cnt  Homozygous Fam Cnt  \\\n",
       "2542     PATHOGENIC          10         8535                   0   \n",
       "4033            VUS           1         8544                   0   \n",
       "\n",
       "      Hemizygous Fam Cnt  Heterozygous Fam Cnt               ref  alt chrom  \n",
       "2542                   0                    10  no alleles found  NaN    13  \n",
       "4033                   0                     1                 T    C    13  \n",
       "\n",
       "[2 rows x 21 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[ df['Genomic Pos (hg19) ']==1435]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p.</th>\n",
       "      <th>Ref (single-base)</th>\n",
       "      <th>Alt (single-base)</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: G&gt;A,G&gt;T,G&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: C&gt;G,C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>S1282_G1285del</td>\n",
       "      <td>multiple alleles found: CTGCCAAGGA&gt;G,CTGCCAAGG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTGCCAAGG</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>S1282_G1285delinsR</td>\n",
       "      <td>multiple alleles found: CTGCCAAGGA&gt;G,CTGCCAAGG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTGCCAAGG</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: G&gt;C,G&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1026</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: C&gt;A,C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1256</th>\n",
       "      <td>H1003LfsX29</td>\n",
       "      <td>multiple alleles found: A&gt;-,A&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1952</th>\n",
       "      <td>Q52_Q55del</td>\n",
       "      <td>multiple alleles found: -&gt;GCAGCAGCAGCA,GCAGCAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCAGCAGCAGCA</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953</th>\n",
       "      <td>Q51_Q55del</td>\n",
       "      <td>multiple alleles found: -&gt;GCAGCAGCAGCAGCA,GCAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCAGCAGCAGCAGCA</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2392</th>\n",
       "      <td>Q70_P78dup</td>\n",
       "      <td>multiple alleles found: AGCAGCAGCAGCCGCCGCCGCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>AGCAGCAGCAGCCGCCGCCGCCGCCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>A62_P69del</td>\n",
       "      <td>multiple alleles found: CGCCCCGCAACCGCCGCCGCCG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CGCCCCGCAACCGCCGCCGCC</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2748</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: T&gt;A,T&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3117</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: T&gt;C,T&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3237</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: A&gt;G,A&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3659</th>\n",
       "      <td>G22_G26del</td>\n",
       "      <td>multiple alleles found: GCGGCGGCGGCGGCG&gt;-,GGCG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGGCGGCGGCGGC</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3667</th>\n",
       "      <td>A148_A155dup</td>\n",
       "      <td>multiple alleles found: -&gt;AGCCGCGGCCGCGGCCGCCG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGCCGCGGCCGCGGCCGCCGCGGC</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3668</th>\n",
       "      <td>A109_A115dup</td>\n",
       "      <td>multiple alleles found: -&gt;GGCGGCGGCGGCGGCGGCGG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGGCGGCGGCGGCGGCGGC</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3723</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: T&gt;A,T&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4008</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: -&gt;GG,T&gt;G,A&gt;T,C&gt;G,A&gt;G,G...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4009</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: A&gt;G,G&gt;T,A&gt;C,T&gt;C,C&gt;A,C&gt;...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AGTTCTCAGCTCTTCTAGTCTGGGG</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4010</th>\n",
       "      <td>NaN</td>\n",
       "      <td>multiple alleles found: C&gt;T,T&gt;C,A&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      p.                                  Ref (single-base)  \\\n",
       "221                  NaN                multiple alleles found: G>A,G>T,G>C   \n",
       "297                  NaN                    multiple alleles found: C>G,C>T   \n",
       "298       S1282_G1285del  multiple alleles found: CTGCCAAGGA>G,CTGCCAAGG...   \n",
       "299   S1282_G1285delinsR  multiple alleles found: CTGCCAAGGA>G,CTGCCAAGG...   \n",
       "300                  NaN                    multiple alleles found: G>C,G>T   \n",
       "1026                 NaN                    multiple alleles found: C>A,C>T   \n",
       "1256         H1003LfsX29                    multiple alleles found: A>-,A>G   \n",
       "1952          Q52_Q55del  multiple alleles found: ->GCAGCAGCAGCA,GCAGCAG...   \n",
       "1953          Q51_Q55del  multiple alleles found: ->GCAGCAGCAGCAGCA,GCAG...   \n",
       "2392          Q70_P78dup  multiple alleles found: AGCAGCAGCAGCCGCCGCCGCC...   \n",
       "2393          A62_P69del  multiple alleles found: CGCCCCGCAACCGCCGCCGCCG...   \n",
       "2748                 NaN                    multiple alleles found: T>A,T>G   \n",
       "3117                 NaN                    multiple alleles found: T>C,T>G   \n",
       "3237                 NaN                    multiple alleles found: A>G,A>T   \n",
       "3659          G22_G26del  multiple alleles found: GCGGCGGCGGCGGCG>-,GGCG...   \n",
       "3667        A148_A155dup  multiple alleles found: ->AGCCGCGGCCGCGGCCGCCG...   \n",
       "3668        A109_A115dup  multiple alleles found: ->GGCGGCGGCGGCGGCGGCGG...   \n",
       "3723                 NaN                    multiple alleles found: T>A,T>C   \n",
       "4008                 NaN  multiple alleles found: ->GG,T>G,A>T,C>G,A>G,G...   \n",
       "4009                 NaN  multiple alleles found: A>G,G>T,A>C,T>C,C>A,C>...   \n",
       "4010                 NaN                multiple alleles found: C>T,T>C,A>G   \n",
       "\n",
       "     Alt (single-base)                        ref                          alt  \n",
       "221                NaN                          G                            C  \n",
       "297                NaN                          C                            T  \n",
       "298                NaN                  CTGCCAAGG                            G  \n",
       "299                NaN                  CTGCCAAGG                            G  \n",
       "300                NaN                          G                            T  \n",
       "1026               NaN                          C                            T  \n",
       "1256               NaN                          A                            G  \n",
       "1952               NaN               GCAGCAGCAGCA                            -  \n",
       "1953               NaN            GCAGCAGCAGCAGCA                            -  \n",
       "2392               NaN                          -  AGCAGCAGCAGCCGCCGCCGCCGCCGC  \n",
       "2393               NaN      CGCCCCGCAACCGCCGCCGCC                            -  \n",
       "2748               NaN                          T                            G  \n",
       "3117               NaN                          T                            G  \n",
       "3237               NaN                          A                            T  \n",
       "3659               NaN            GGCGGCGGCGGCGGC                            -  \n",
       "3667               NaN   AGCCGCGGCCGCGGCCGCCGCGGC                            -  \n",
       "3668               NaN      GGCGGCGGCGGCGGCGGCGGC                            -  \n",
       "3723               NaN                          T                            C  \n",
       "4008               NaN                          C                            A  \n",
       "4009               NaN  AGTTCTCAGCTCTTCTAGTCTGGGG                            -  \n",
       "4010               NaN                          A                            G  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crit = df.apply(lambda row: ',' in row['Ref (single-base)'], axis=1)\n",
    "uc = ['p.', 'Ref (single-base)', 'Alt (single-base)', 'ref', 'alt']\n",
    "df[crit][uc]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>p.</th>\n",
       "      <th>c.</th>\n",
       "      <th>Genomic Pos (hg19)</th>\n",
       "      <th>Ref (single-base)</th>\n",
       "      <th>Alt (single-base)</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Exon (Gene)</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>test/Indication</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Pos Fam Cnt</th>\n",
       "      <th>Neg Fam Cnt</th>\n",
       "      <th>Homozygous Fam Cnt</th>\n",
       "      <th>Hemizygous Fam Cnt</th>\n",
       "      <th>Heterozygous Fam Cnt</th>\n",
       "      <th>ref</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22</td>\n",
       "      <td>A206S</td>\n",
       "      <td>c.616</td>\n",
       "      <td>12998</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>GCC</td>\n",
       "      <td>TCC</td>\n",
       "      <td>exn5-39</td>\n",
       "      <td>ADSL</td>\n",
       "      <td>Comp_epi</td>\n",
       "      <td>NM_000026.2</td>\n",
       "      <td>VUS</td>\n",
       "      <td>8</td>\n",
       "      <td>8275</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22</td>\n",
       "      <td>A260V</td>\n",
       "      <td>c.779</td>\n",
       "      <td>14480</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GCA</td>\n",
       "      <td>GTA</td>\n",
       "      <td>exn7-14</td>\n",
       "      <td>ADSL</td>\n",
       "      <td>Comp_epi</td>\n",
       "      <td>NM_000026.2</td>\n",
       "      <td>VUS</td>\n",
       "      <td>1</td>\n",
       "      <td>8282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr22</td>\n",
       "      <td>A2V</td>\n",
       "      <td>c.5</td>\n",
       "      <td>564</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>GCG</td>\n",
       "      <td>GTG</td>\n",
       "      <td>exn1+64</td>\n",
       "      <td>ADSL</td>\n",
       "      <td>Infant_epi</td>\n",
       "      <td>NM_000026.2</td>\n",
       "      <td>PATHOGENIC</td>\n",
       "      <td>1</td>\n",
       "      <td>8282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22</td>\n",
       "      <td>A480G</td>\n",
       "      <td>c.1439</td>\n",
       "      <td>20507</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>GCA</td>\n",
       "      <td>GGA</td>\n",
       "      <td>exn13-68</td>\n",
       "      <td>ADSL</td>\n",
       "      <td>Infant_epi</td>\n",
       "      <td>NM_000026.2</td>\n",
       "      <td>VUS</td>\n",
       "      <td>1</td>\n",
       "      <td>8282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22</td>\n",
       "      <td>C483R</td>\n",
       "      <td>c.1447</td>\n",
       "      <td>20515</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>TGT</td>\n",
       "      <td>CGT</td>\n",
       "      <td>exn13-60</td>\n",
       "      <td>ADSL</td>\n",
       "      <td>Child_epi</td>\n",
       "      <td>NM_000026.2</td>\n",
       "      <td>VUS</td>\n",
       "      <td>1</td>\n",
       "      <td>8282</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr     p.      c. Genomic Pos (hg19)  Ref (single-base)  \\\n",
       "0  chr22  A206S   c.616               12998                 G   \n",
       "1  chr22  A260V   c.779               14480                 C   \n",
       "2  chr22    A2V     c.5                 564                 C   \n",
       "3  chr22  A480G  c.1439               20507                 C   \n",
       "4  chr22  C483R  c.1447               20515                 T   \n",
       "\n",
       "  Alt (single-base)  Ref  Alt Exon (Gene) Gene Symbol test/Indication  \\\n",
       "0                 T  GCC  TCC     exn5-39        ADSL        Comp_epi   \n",
       "1                 T  GCA  GTA     exn7-14        ADSL        Comp_epi   \n",
       "2                 T  GCG  GTG     exn1+64        ADSL      Infant_epi   \n",
       "3                 G  GCA  GGA    exn13-68        ADSL      Infant_epi   \n",
       "4                 C  TGT  CGT    exn13-60        ADSL       Child_epi   \n",
       "\n",
       "    Transcript Classification  Pos Fam Cnt  Neg Fam Cnt  Homozygous Fam Cnt  \\\n",
       "0  NM_000026.2            VUS            8         8275                   0   \n",
       "1  NM_000026.2            VUS            1         8282                   0   \n",
       "2  NM_000026.2     PATHOGENIC            1         8282                   0   \n",
       "3  NM_000026.2            VUS            1         8282                   0   \n",
       "4  NM_000026.2            VUS            1         8282                   0   \n",
       "\n",
       "   Hemizygous Fam Cnt  Heterozygous Fam Cnt ref  \n",
       "0                   0                     8   G  \n",
       "1                   0                     1   C  \n",
       "2                   0                     1   C  \n",
       "3                   0                     1   C  \n",
       "4                   0                     1   T  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chr                                                                 chr15\n",
       "p.                                                             Q51_Q55del\n",
       "c.                                                              c.144_158\n",
       "Genomic Pos (hg19)                                                   1685\n",
       "Ref (single-base)       multiple alleles found: ->GCAGCAGCAGCAGCA,GCAG...\n",
       "Alt (single-base)                                                     NaN\n",
       "Ref                                                                   NaN\n",
       "Alt                                                                   NaN\n",
       "Exon (Gene)                                                           NaN\n",
       "Gene Symbol                                                          POLG\n",
       "test/Indication                                                  Comp_epi\n",
       "Transcript                                                    NM_002693.2\n",
       "Classification                                              LIKELY BENIGN\n",
       "Pos Fam Cnt                                                             1\n",
       "Neg Fam Cnt                                                          8419\n",
       "Homozygous Fam Cnt                                                      0\n",
       "Hemizygous Fam Cnt                                                      0\n",
       "Heterozygous Fam Cnt                                                    1\n",
       "ref                                                       GCAGCAGCAGCAGCA\n",
       "Name: 1953, dtype: object"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[1953]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
